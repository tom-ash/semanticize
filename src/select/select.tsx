import React, { useMemo, useState } from "react";
import { TextInput } from "../text-input/text-input";

type Value = string | number | null;
type Text = string | undefined;

interface Option {
  value: Value;
  text: Text;
}

interface SelectInterface {
  (props: {
    id?: string;
    className?: string;
    label?: string | React.ReactElement;
    options: Option[];
    value: Value;
    onSelect(value: Value): void;
  }): React.ReactElement;
}

export const Select: SelectInterface = props => {
  const { id, className, label, options, value, onSelect } = props;
  const [isFocused, changeIsFocused] = useState(false);
  const ref = React.useRef<HTMLInputElement>(null);

  const [isCurrent, changeIsCurrent] = useState(0);

  const onKeyDown = (e: React.KeyboardEvent) => {
    const key = e.key;

    if (key === "ArrowDown") {
      const newIsCurrent = isCurrent + 1;

      if (newIsCurrent >= options.length) {
        changeIsCurrent(0);
      } else {
        changeIsCurrent(newIsCurrent);
      }
    } else if (key === "ArrowUp") {
      const newIsCurrent = isCurrent - 1;

      if (newIsCurrent < 0) {
        changeIsCurrent(options.length - 1);
      } else {
        changeIsCurrent(newIsCurrent);
      }
    } else if (key === "Enter") {
      const currentOptionValue = options[isCurrent].value;

      onSelect(currentOptionValue);
      ref.current?.blur();
    }
  };

  const currentOption = useMemo(
    () => options.find((option) => {
      return option.value === value
    }),
    [options, value]
  );

  return (
    <TextInput
      ref={ref}
      id={id}
      className={className}
      label={label}
      onFocus={() => changeIsFocused(true)}
      onBlur={() => {
        changeIsFocused(false);
      }}
      value={currentOption?.text || ''}
      disableCaret={true}
      onKeyDown={onKeyDown}
    >
      {isFocused && (
        <ul>
          {options.map((option, index) => {
            const { value, text } = option;
            const className = isCurrent === index ? "current" : undefined;

            return (
              <li
                key={value}
                className={className}
                onMouseDown={e => {
                  e.preventDefault();
                }}
                onClick={() => {
                  onSelect(value);
                  ref.current?.blur();
                }}
              >
                {text}
              </li>
            );
          })}
        </ul>
      )}
    </TextInput>
  );
};
