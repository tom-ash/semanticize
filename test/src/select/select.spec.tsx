import React from "react";
import { describe, expect, test } from "@jest/globals";
import { Select } from "../../../src/select/select";
import renderer, { act } from "react-test-renderer";

const options = [
  { value: "foo", text: "Foo" },
  { value: "bar", text: "Bar" },
  { value: "baz", text: "Baz" },
];

const onSelect = value => console.log(value);

const requiredProps = {
  options,
  value: "foo",
  onSelect,
};

describe("Select", () => {
  describe("label", () => {
    describe("when the provided label is string", () => {
      test("renders with the provided label", () => {
        const component = renderer.create(
          <Select
            label="Test"
            {...requiredProps}
          />
        );

        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
      });
    });

    describe("when the provided label is ReactElement", () => {
      test("renders with the provided label", () => {
        const component = renderer.create(
          <Select
            label="Test"
            {...requiredProps}
          />
        );

        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
      });
    });

    describe("when no label is provided", () => {
      test("renders without a label", () => {
        const component = renderer.create(<Select {...requiredProps} />);

        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
      });
    });
  });

  describe("options", () => {
    describe("when the select's text input is not focused", () => {
      test("does not show options", () => {
        const component = renderer.create(
          <Select
            label="Test"
            {...requiredProps}
          />
        );

        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
      });
    });

    describe("when the select's text input is focused", () => {
      test("shows options", () => {
        const component = renderer.create(
          <Select
            label="Test"
            {...requiredProps}
          />
        );

        act(() => {
          component.root.findByType("input").props.onFocus({ target: { value: "" } });
        });

        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
      });
    });
  });

  describe("value", () => {
    describe("when an option is selected", () => {
      test("it sets the text input value to option text", () => {
        const component = renderer.create(
          <Select
            label="Test"
            {...requiredProps}
          />
        );

        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
      });
    });

    describe("when an option is not selected", () => {
      test("it sets the text input value to an empty string", () => {
        const component = renderer.create(
          <Select
            label="Test"
            {...{ ...requiredProps, value: null }}
          />
        );

        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
      });
    });
  });

  describe("children", () => {
    describe("when children are not provided", () => {
      test("it renders the component without children", () => {
        const component = renderer.create(
          <Select
            label="Test"
            {...requiredProps}
          />
        );

        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
      });
    });

    describe("when children are provided", () => {
      test("it renders the component with children", () => {
        const component = renderer.create(
          <Select
            label="Test"
            children={<>TEST</>}
            {...{ ...requiredProps, value: null }}
          />
        );

        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
      });
    });
  });
});
