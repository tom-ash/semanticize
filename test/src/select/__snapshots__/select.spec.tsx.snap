// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select children when children are not provided it renders the component without children 1`] = `
<div
  className=""
>
  <label>
    Test
  </label>
  <input
    onBlur={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    style={
      {
        "caretColor": "transparent",
      }
    }
    type="text"
    value="Foo"
  />
</div>
`;

exports[`Select children when children are provided it renders the component with children 1`] = `
<div
  className=""
>
  <label>
    Test
  </label>
  <input
    onBlur={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    style={
      {
        "caretColor": "transparent",
      }
    }
    type="text"
    value=""
  />
  TEST
</div>
`;

exports[`Select label when no label is provided renders without a label 1`] = `
<div
  className=""
>
  <input
    onBlur={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    style={
      {
        "caretColor": "transparent",
      }
    }
    type="text"
    value="Foo"
  />
</div>
`;

exports[`Select label when the provided label is ReactElement renders with the provided label 1`] = `
<div
  className=""
>
  <label>
    Test
  </label>
  <input
    onBlur={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    style={
      {
        "caretColor": "transparent",
      }
    }
    type="text"
    value="Foo"
  />
</div>
`;

exports[`Select label when the provided label is string renders with the provided label 1`] = `
<div
  className=""
>
  <label>
    Test
  </label>
  <input
    onBlur={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    style={
      {
        "caretColor": "transparent",
      }
    }
    type="text"
    value="Foo"
  />
</div>
`;

exports[`Select options when the select's text input is focused shows options 1`] = `
<div
  className=""
>
  <label>
    Test
  </label>
  <input
    onBlur={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    style={
      {
        "caretColor": "transparent",
      }
    }
    type="text"
    value="Foo"
  />
  <ul>
    <li
      className="current"
      onClick={[Function]}
      onMouseDown={[Function]}
    >
      Foo
    </li>
    <li
      onClick={[Function]}
      onMouseDown={[Function]}
    >
      Bar
    </li>
    <li
      onClick={[Function]}
      onMouseDown={[Function]}
    >
      Baz
    </li>
  </ul>
</div>
`;

exports[`Select options when the select's text input is not focused does not show options 1`] = `
<div
  className=""
>
  <label>
    Test
  </label>
  <input
    onBlur={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    style={
      {
        "caretColor": "transparent",
      }
    }
    type="text"
    value="Foo"
  />
</div>
`;

exports[`Select value when an option is not selected it sets the text input value to an empty string 1`] = `
<div
  className=""
>
  <label>
    Test
  </label>
  <input
    onBlur={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    style={
      {
        "caretColor": "transparent",
      }
    }
    type="text"
    value=""
  />
</div>
`;

exports[`Select value when an option is selected it sets the text input value to option text 1`] = `
<div
  className=""
>
  <label>
    Test
  </label>
  <input
    onBlur={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    style={
      {
        "caretColor": "transparent",
      }
    }
    type="text"
    value="Foo"
  />
</div>
`;
